linkml/generators/common/name_compatibility.py

```diff
@@ -0,0 +1,84 @@
+
+    # heading double underscores and digit reserved to names starting with a digit
+    re_reserved_heading_digit = re.compile("^__[0-9]")
+    # valid name between double underscores is reserved for unicode name transformations
+    re_reserved_unicode_name_transformation = re.compile("__[0-9a-zA-Z][0-9a-zA-Z_]*__")
+    # something like '__U_xxxx_' is reserved for unicode code transformations
+    re_reserved_unicode_code_transformation = re.compile("__U_[0-9a-eA-E]{4}_")
+    # strings starting with a digit
+    re_heading_digit = re.compile("^[0-9]")
+    # character that is neither alphanumeric nor underscore
+    re_no_alphanum_underscore = re.compile("[^0-9a-zA-Z_]")
+
+    def __init__(self, profile: Literal["graphql"], do_not_fix: bool = False):
+        """Specify the naming policy on instantiation"""
+        self.profile = profile
+        self.do_not_fix = do_not_fix
+
+    def _no_heading_digits(self, input: str) -> str:
+        """Ensure name does not start with a heading digit"""
+        output = input
+        if self.re_heading_digit.match(input):
+            if self.do_not_fix:
+                raise ValueError(f"Name '{input}' starts with digit (illegal GraphQL) and will not be fixed!")
+            output = f"__{input}"
+        return output
+
+    def _transform_char(self, char: str) -> str:
+        """Transform unsupported characters"""
+        if len(char) != 1:
+            raise Exception(f"Single character expected, but got '{char}'!!")
+        # try to use names for ASCII characters
+        if ord(char) < 128:
+            try:
+                # unicodedata.lookup should be able to invert the transformation
+                return f"__{unicodedata.name(char).replace(' ','_')}__"
+            except ValueError:
+                pass
+        # fallback to code-transformation if none of the previous has worked
+        return f"__U_{ord(char):04X}_"
+
+    def _only_alphanum_underscore(self, input: str) -> str:
+        """Ensure name does not contain any unsupported characters"""
+        output = input
+        # with re.split and re.findall we get in the same order and separated in two arrays
+        # the substrings between special characters and the special characters
+        no_alphanum_underscore_match = self.re_no_alphanum_underscore.findall(input)
+        if no_alphanum_underscore_match:
+            if self.do_not_fix:
+                raise ValueError(f"Name '{input}' contains a character illegal in GraphQL and will not be fixed!")
+            to_keep = self.re_no_alphanum_underscore.split(input)
+            # first comes first substring to keep
+            output = to_keep[0]
+            # each char replacement is followed by the next substring to keep
+            for offset in range(0, len(to_keep) - 1):
+                output = output + self._transform_char(no_alphanum_underscore_match[offset])
+                output = output + to_keep[offset + 1]
+        return output
```

Comment:
I think your code would not do what is needed. I might not understand your code right, though...

What is needed is a "split" of substrings to be kept unmodified and characters to be transformed. The special characters acts somehow as separators. You them in two separate arrays (`re.split` and `re.findall`). Finally they need to be "zipped" back together: one substring to be kept, one transformed character, one substring to be kept, one transformed character and so on and so forth.

In any case, it would be only an optimization that can be applied later on. So resolving this conversation. Feel free to propose the optimization on a separate PR.

Reply:
I think I miswrote the last line and it should be. .sub, but re.sub does all that - you dont need to split, fix, and re-zip, just match, apply function to match, done.